### PR TITLE
[Flang][MLIR][OpenMP] Create a deferred declare target marking process for Bridge.cpp

### PR DESCRIPTION
**Pull request by agozillon**
This patch seeks to create a process that happens on module finalization for OpenMP, in which a list of operations that had declare target directives applied to them and were not generated at the time of processing the original declare target directive are re-checked to apply the appropriate declare target semantics.

This works by maintaining a vector of declare target related data inside of the FIR converter, in this case the symbol and the two relevant unsigned integers representing the enumerators. This vector is added to via a new function called from Bridge.cpp, insertDeferredDeclareTargets, which happens prior to the processing of the directive (similarly to getDeclareTargetFunctionDevice currently for requires), it effectively checks if the Operation the declare target directive is applied to currently exists, if it doesn't it appends to the vector. This is a seperate function to the processing of the declare target via the overloaded genOMP as we unfortunately do not have access to the list without passing it through every call, as the AbstractConverter we pass will not allow access to it (I've seen no other cases of casting it to a FirConverter, so I opted to not do that).

The list is then processed at the end of the module in the finalizeOpenMPLowering function in Bridge by calling a new function markDelayedDeclareTargetFunctions which marks the latently generated operations. In certain cases, some still will not be generated, e.g. if an interface is defined, marked as declare target, but has no definition or usage in the module then it will not be emitted to the module, so due to these cases we must silently ignore when an operation has not been found via it's symbol.

The main use-case for this (although, I imagine there is others) is for processing interfaces that have been declared in a module with a declare target directive but do not have their implementation defined in the same module. For example, inside of a seperate C++ module that will be linked in. In cases where the interface is called inside of a target region it'll be marked as used on device appropriately (although, realistically a user should explicitly mark it to match the corresponding definition), however, in cases where it's used in a non-clear manner through something like a function pointer passed to an external call we require this explicit marking, which this patch adds support for (currently will cause the compiler to crash).

An alternative to this approach may be to do it in the implicit marking pass, by searching for function pointers/addresses used in target regions and marking the FuncOp's appropriately. But this doesn't seem like the right approach, we would be ignoring the users explicit markings at that point if they exist (we'd have no way of saving the data and transferring it to the pass trivially) in lieu of our own assumptions, and this would also not support other possible operation types that may need deferred marking.

### CODE REVIEW

**Review by ergawy — COMMENTED** [review]
Still working on understanding the details of changes introduced in the PR. But for now just a few quick remarks from an initial look at the first few files.

File: flang/include/flang/Lower/OpenMP.h
Comment (agozillon) [reply flang/include/flang/Lower/OpenMP.h:102]: Not sure what you mean by this unfortunately, if you could elaborate more that'd be appreciated!

File: flang/include/flang/Lower/OpenMP.h
Comment (agozillon) [reply flang/include/flang/Lower/OpenMP.h:96]: Good idea, happy to do a type alias, however, I'd lean towards keeping it a tuple as that seems to be the standard in most cases, but I'm not overly bothered either way, so happy to change it if you feel strongly about it!

**Comment by agozillon** [comment]
Small ping for some reviewer attention next week if any time can be spared

**Comment by agozillon** [comment]
Another small ping for reviewer attention if at all possible before I go ahead and address the existing comments. 

I am primarily interested in if this seems like a reasonable approach to the problem from @kiranchandramohan or @jeanPerier and if not possible suggestions towards an alternative!

File: flang/lib/Lower/Bridge.cpp
Comment (clementval) [inline flang/lib/Lower/Bridge.cpp:5037]: Can the name be more explicit with an omp prefix of smth?

Comment (agozillon) [reply flang/lib/Lower/Bridge.cpp:5037]: It can indeed, I'll also try and make the previously suggested typedef more OpenMP explicit as well.

**Comment by agozillon** [comment]
Updated the PR based on reviewer comments [(commit)](https://github.com/llvm/llvm-project/pull/78502/commits/3feda65753323823e226f9eda0ec25868f8d93e1#diff-0f70165af00b05acb5abeb11198079b793a20acf8f069edd2dbe7b34ac67b328). 

I rebased as it's been a little while since the PR has been updated fully, a side affect of that is the need to force push, so I apologies to reviewers if that makes things difficult when reviewing the recent changes. 

The recent changes encompass some minor changes: 

- Moved to using a type-alias for the tuple
- Made names a little more consistent and OpenMP specific across the board

**Comment by agozillon** [comment]
Small ping to see if we can move this PR forward one way or another! Thank you very much ahead of time for any of your precious time you can give to the review! :-)

**Comment by kiranchandramohan** [comment]
@agozillon This looks OK, but the code that handles `declare target` has changed quite a bit after the initial patch. Since it is split across different places (lowering, pass, translation to LLVM), it would be good to document this.

**Comment by agozillon** [comment]
Thank you Kiran, I'll perform a rebase and add a document of the overall process as I understand it now, and @skatrak who has also worked on it can verify if I am incorrect in any places when it's up and he has time!

**Comment by agozillon** [comment]
Recent commit performed a rebase (nothing too interesting change wise, just re-aligning certain function calls with some recent changes).

It also added new documentation on declare target handling, I would appreciate if you could have a little look at the parts that you've worked on @skatrak to make sure I've captured the essence of it!

**Comment by agozillon** [comment]
Most recent PR (hopefully) fixes the documentation build bug I introduced:

* Actually add the document to the index file, which I forgot to do
* change the MLIR code block into a blank code block in the new MD as it's unsupported at the moment it seems

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:100]:
```suggestion
For `GlobalOp`'s, the processing is a little more complex. We currently leverage the `registerTargetGlobalVariable` and `getAddrOfDeclareTargetVar` `OMPIRBuilder` functions shared with Clang. These two functions invoke each other depending on the clauses and options provided to the `OMPIRBuilder` (in particular, unified shared memory). Their main purposes are the generation of a new global device pointer with a "ref_" prefix on the device and enqueuing metadata generation by the `OMPIRBuilder` to be produced at module finalization time. This is done for both host and device and it links the newly generated device global pointer and the host pointer together across the two modules. 
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:102]:
```suggestion
Similarly to other metadata (e.g. for `TargetOp`) that is shared across both host and device modules, processing of `GlobalOp`'s in the device needs access to the previously generated host IR file, which is done through another `attribute` applied to the `ModuleOp` by the compiler frontend. The file is loaded in and consumed by the `OMPIRBuilder` to populate it's `OffloadInfoManager` data structures, keeping host and device appropriately synchronised.
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:124]:
```suggestion
* Marking global variables with `declare target`'s `to` clause. A lot of the lowering should exist, but it needs further testing and likely some further changes to fully function.
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:116]:
```suggestion
   environment utilising `declare target`. This may not work for all types yet, but for scalars and arrays
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:117]:
```suggestion
   of scalars, it should.
```

File: flang/docs/index.md
Comment (skatrak) [inline flang/docs/index.md:72]: Nit: Move one line up for alphabetical order.

**Review by skatrak — COMMENTED** [review]
Thank you Andrew for this work. I think the overall approach makes sense, my comments are just over a few details.

I added quite a few suggestions to the documentation, to try to make it a bit easier to follow and perhaps clarify a couple of points. Feel free to ignore any suggestions there that you think don't improve the text.

File: flang/include/flang/Lower/OpenMP.h
Comment (agozillon) [reply flang/include/flang/Lower/OpenMP.h:58]: Can try to do, it was intended to mimic the existing DeclareTargetCapturePair we have already, I can maybe use the dialect enums, it depends how bothered we are about including the dialect header or forward declaring them (i recall forward declaring being a little bit awkward, but it's been a while)

File: flang/lib/Lower/Bridge.cpp
Comment (agozillon) [reply flang/lib/Lower/Bridge.cpp:5075]: Will do it as part of this PR as it's a small change and makes swapping the contents of this patch to a ref a little simpler (we get the symbol from the pair, which isn't a problem when taking a copy, but is if we take a ref currently as I believe the copy in the pair falls out of scope before we process the deferred values, swapping both to ref means we're pointing at the original which persists to and past when we require it)

File: flang/lib/Lower/OpenMP.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP.cpp:4020]: I think you've corrected me on this exact same spelling mistake a few times now... thank you very much :D might be time to integrate a spell checker into my IDE... especially for the documentation component

**Comment by agozillon** [comment]
Thank you Sergio for the feedback! I believe I've addressed it all in the recent commit, which should: 
> Fix up the documentation with recommended changes
> Fix alphabetical ordering of the index
> Change to a struct from a tuple
> Add a TODO message

**Comment by agozillon** [comment]
Performed a rebase on recent refactoring to the OpenMP lowering, nothing too drastic required in this PR it is mostly the same!

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:111]:
```suggestion
`gatherOpenMPDeferredDeclareTargets`, which gathers information relevant to the
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:128]:
```suggestion
`markOpenMPDeferredDeclareTargetFunctions`, which is called from the lowering
```

File: flang/lib/Lower/Bridge.cpp
Comment (skatrak) [inline flang/lib/Lower/Bridge.cpp:5095]: I think we should be able to use the default now.
```suggestion
  llvm::SmallVector<Fortran::lower::OMPDeferredDeclareTargetInfo>
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:65]:
```suggestion
the lowering can use in different ways as is necessary. 
```

File: flang/docs/OpenMP-declare-target.md
Comment (skatrak) [inline flang/docs/OpenMP-declare-target.md:227]:
```suggestion
To appropriately delete these no longer necessary symbols we would need a 
```

**Review by skatrak — COMMENTED** [review]
Thank you for addressing my comments Andrew, I've got a few nits left after a second look, but they should be trivial and this looks nearly ready to me.

I marked as resolved my prior comments that have been addressed, but there are a couple that I left open because they don't seem to have been incorporated yet. Let me know if this is just because you don't think these documentation suggestions improve readability.

**Review by skatrak — APPROVED** [review]
Thank you Andrew for addressing my comments. This LGTM, but leave some time before merging in case @ergawy, @kiranchandramohan or someone else has any further suggestions.

**Comment by agozillon** [comment]
Thank you very much Sergio for the detailed review, I appreciate it greatly :-) as well as @ergawy's, I think the PR is in much better shape now. 

More than happy to wait a little bit, it would be nice to get one more approval on the PR before landing!

**Review by kiranchandramohan — APPROVED** [review]
LG

**Comment by agozillon** [comment]
Thank you very much @kiranchandramohan ! I'll land the PR just now then as it's passed the CI and has no conflicts.

As for the presentation I'd be happy to do it at some point, although, I am admittedly not the best presenter! Myself and @skatrak  can likely discuss who would prefer to do it in the future offline.

**Pull request merged**
